File: scripts/robust_control/04_robust_control/generate_results.m
root = fileparts(fileparts(fileparts(fileparts(mfilename('fullpath')))));
fig_dir = fullfile(root, 'figures', '04_robust_control');
gen_dir = fullfile(root, 'generated', '04_robust_control');

if ~exist(fig_dir, 'dir')
    mkdir(fig_dir);
end
if ~exist(gen_dir, 'dir')
    mkdir(gen_dir);
end

B = [0; 1];
D = [0; 1];
E = diag([1.0, 0.4]);
Q = diag([6.0, 2.0]);
R = 1.0;

A_nominal = uncertainty_matrix(0.0, 0.0);
[P, ~, K_lqr] = care(A_nominal, B, Q, R);
K = -K_lqr;

rho_1_grid = linspace(-0.6, 0.6, 81);
rho_2_grid = linspace(-0.25, 0.25, 81);
open_loop_alpha = zeros(length(rho_2_grid), length(rho_1_grid));
closed_loop_alpha = zeros(length(rho_2_grid), length(rho_1_grid));

for i = 1:length(rho_2_grid)
    for j = 1:length(rho_1_grid)
        A = uncertainty_matrix(rho_1_grid(j), rho_2_grid(i));
        A_cl = A + B * K;
        open_loop_alpha(i, j) = max(real(eig(A)));
        closed_loop_alpha(i, j) = max(real(eig(A_cl)));
    end
end

vertex_points = {
    'V1', -0.6, -0.25;
    'V2', -0.6, 0.25;
    'V3',  0.6, -0.25;
    'V4',  0.6, 0.25;
    'Nominal', 0.0, 0.0
};
frequency_grid = logspace(-3, 3, 4000);
vertex_metrics = zeros(size(vertex_points, 1), 6);
vertex_labels = strings(size(vertex_points, 1), 1);

for i = 1:size(vertex_points, 1)
    vertex_labels(i) = string(vertex_points{i, 1});
    rho_1 = vertex_points{i, 2};
    rho_2 = vertex_points{i, 3};
    A = uncertainty_matrix(rho_1, rho_2);
    A_cl = A + B * K;
    residual = A_cl' * P + P * A_cl;
    vertex_metrics(i, :) = [
        rho_1, ...
        rho_2, ...
        max(real(eig(A))), ...
        max(real(eig(A_cl))), ...
        max(real(eig(residual))), ...
        estimate_hinf_norm(A_cl, D, E, frequency_grid)
    ];
end

representative_cases = {
    'rho1=-0.6, rho2=-0.25', -0.6, -0.25;
    'rho1=0.0, rho2=0.0', 0.0, 0.0;
    'rho1=0.6, rho2=0.25', 0.6, 0.25
};
case_count = size(representative_cases, 1);
response_data = cell(case_count, 4);
peak_table = zeros(case_count, 2);

for i = 1:case_count
    label = string(representative_cases{i, 1});
    rho_1 = representative_cases{i, 2};
    rho_2 = representative_cases{i, 3};
    A_cl = uncertainty_matrix(rho_1, rho_2) + B * K;
    [time, states] = ode45(@(t, x) A_cl * x + [0; disturbance_signal(t)], [0, 15], [0; 0]);
    controls = (K * states')';
    response_data{i, 1} = label;
    response_data{i, 2} = time;
    response_data{i, 3} = states;
    response_data{i, 4} = controls;
    peak_table(i, :) = [max(abs(states), [], 'all'), max(abs(controls), [], 'all')];
end

summary.A_nominal = A_nominal;
summary.B = B;
summary.D = D;
summary.E = E;
summary.K = K;
summary.P = P;
summary.uncertainty_box.rho_1 = [-0.6, 0.6];
summary.uncertainty_box.rho_2 = [-0.25, 0.25];
summary.worst_closed_loop_alpha_on_grid = max(closed_loop_alpha, [], 'all');
summary.best_closed_loop_alpha_on_grid = min(closed_loop_alpha, [], 'all');
summary.worst_open_loop_alpha_on_grid = max(open_loop_alpha, [], 'all');
summary.best_open_loop_alpha_on_grid = min(open_loop_alpha, [], 'all');
summary.vertex_labels = vertex_labels;
summary.vertex_metrics = vertex_metrics;
summary.representative_case_labels = string(representative_cases(:, 1));
summary.representative_peaks = peak_table;

summary_json = jsonencode(summary, PrettyPrint=true);
fid = fopen(fullfile(gen_dir, 'summary.json'), 'w');
fwrite(fid, summary_json, 'char');
fclose(fid);

metric_headers = {'label', 'rho_1', 'rho_2', 'open_loop_alpha', 'closed_loop_alpha', ...
    'lyapunov_residual_alpha', 'hinf_estimate'};
fid = fopen(fullfile(gen_dir, 'vertex_metrics.csv'), 'w');
fprintf(fid, '%s,%s,%s,%s,%s,%s,%s\n', metric_headers{:});
for i = 1:size(vertex_metrics, 1)
    fprintf( ...
        fid, ...
        '%s,%.6f,%.6f,%.6f,%.6f,%.6f,%.6f\n', ...
        char(vertex_labels(i)), ...
        vertex_metrics(i, 1), ...
        vertex_metrics(i, 2), ...
        vertex_metrics(i, 3), ...
        vertex_metrics(i, 4), ...
        vertex_metrics(i, 5), ...
        vertex_metrics(i, 6) ...
    );
end
fclose(fid);

figure('Position', [120, 120, 1180, 420], 'Color', 'w');
tiledlayout(1, 2, 'Padding', 'compact', 'TileSpacing', 'compact');

nexttile;
imagesc(rho_1_grid, rho_2_grid, open_loop_alpha);
set(gca, 'YDir', 'normal');
colorbar;
xlabel('\rho_1');
ylabel('\rho_2');
title('Open-loop spectral abscissa');

nexttile;
imagesc(rho_1_grid, rho_2_grid, closed_loop_alpha);
set(gca, 'YDir', 'normal');
colorbar;
xlabel('\rho_1');
ylabel('\rho_2');
title('Closed-loop spectral abscissa');
exportgraphics(gcf, fullfile(fig_dir, 'spectral_abscissa_scan.png'), 'Resolution', 220);
close(gcf);

figure('Position', [150, 150, 840, 420], 'Color', 'w');
bar(categorical(vertex_labels), vertex_metrics(:, 6), 'FaceColor', [0.29, 0.47, 0.69]);
hold on;
yline(0.5, '--', '\gamma = 0.5', 'LineWidth', 1.4, 'Color', [0.74, 0.26, 0.17]);
ylabel('Estimated gain');
title('Estimated H_{\infty} gain at interval vertices');
grid on;
exportgraphics(gcf, fullfile(fig_dir, 'vertex_hinf_estimates.png'), 'Resolution', 220);
close(gcf);

figure('Position', [160, 140, 920, 560], 'Color', 'w');
tiledlayout(2, 1, 'Padding', 'compact', 'TileSpacing', 'compact');

nexttile;
hold on;
for i = 1:case_count
    plot(response_data{i, 2}, response_data{i, 3}(:, 1), 'LineWidth', 1.7, 'DisplayName', response_data{i, 1} + " : x_1");
end
ylabel('x_1(t)');
title('Closed-loop states under decaying disturbance');
grid on;
legend('Location', 'northeastoutside');

nexttile;
hold on;
for i = 1:case_count
    plot(response_data{i, 2}, response_data{i, 3}(:, 2), 'LineWidth', 1.7, 'DisplayName', response_data{i, 1} + " : x_2");
end
xlabel('Time (s)');
ylabel('x_2(t)');
grid on;
legend('Location', 'northeastoutside');
exportgraphics(gcf, fullfile(fig_dir, 'state_response_under_disturbance.png'), 'Resolution', 220);
close(gcf);

figure('Position', [180, 160, 920, 420], 'Color', 'w');
hold on;
for i = 1:case_count
    plot(response_data{i, 2}, response_data{i, 4}, 'LineWidth', 1.8, 'DisplayName', response_data{i, 1});
end
time_plot = linspace(0, 15, 1501);
plot(time_plot, disturbance_signal(time_plot), '--', 'LineWidth', 1.2, 'Color', [0.15, 0.15, 0.15], 'DisplayName', 'w(t)');
xlabel('Time (s)');
ylabel('Signal value');
title('Control input and disturbance');
grid on;
legend('Location', 'northeastoutside');
exportgraphics(gcf, fullfile(fig_dir, 'control_response_under_disturbance.png'), 'Resolution', 220);
close(gcf);

fprintf('Robust control reproduction completed.\n');
fprintf('K = [%.6f %.6f]\n', K(1), K(2));
fprintf('Worst closed-loop spectral abscissa on grid: %.6f\n', max(closed_loop_alpha, [], 'all'));
fprintf('Worst estimated vertex Hinf gain: %.6f\n', max(vertex_metrics(:, 6)));

function A = uncertainty_matrix(rho_1, rho_2)
    A = [0.0, 1.0; 1.2 + rho_1, 0.3 + rho_2];
end

function value = disturbance_signal(t)
    value = 0.8 .* exp(-0.3 .* t) .* sin(2.4 .* t);
end

function gain = estimate_hinf_norm(A_cl, D, E, frequencies)
    gain = 0.0;
    identity = eye(size(A_cl));
    for idx = 1:length(frequencies)
        omega = frequencies(idx);
        transfer = E * ((1i * omega * identity - A_cl) \ D);
        sigma = norm(transfer, 2);
        gain = max(gain, sigma);
    end
end

File: experiments/robust_control/04_robust_control/generate_results.py
from __future__ import annotations

import csv
import json
from pathlib import Path

import matplotlib.pyplot as plt
import numpy as np
from scipy import linalg
from scipy.integrate import solve_ivp


ROOT = Path(__file__).resolve().parents[3]
FIG_DIR = ROOT / "figures" / "04_robust_control"
GEN_DIR = ROOT / "generated" / "04_robust_control"


def uncertainty_matrix(rho_1: float, rho_2: float) -> np.ndarray:
    return np.array([[0.0, 1.0], [1.2 + rho_1, 0.3 + rho_2]], dtype=float)


def disturbance(t: np.ndarray | float) -> np.ndarray | float:
    return 0.8 * np.exp(-0.3 * t) * np.sin(2.4 * t)


def estimate_hinf_norm(
    a_cl: np.ndarray,
    d_mat: np.ndarray,
    e_mat: np.ndarray,
    frequencies: np.ndarray,
) -> float:
    max_sigma = 0.0
    identity = np.eye(a_cl.shape[0], dtype=complex)
    for omega in frequencies:
        transfer = e_mat @ np.linalg.inv(1j * omega * identity - a_cl) @ d_mat
        sigma = np.linalg.norm(transfer, ord=2)
        max_sigma = max(max_sigma, float(sigma))
    return max_sigma


def simulate_case(
    a_cl: np.ndarray,
    k_gain: np.ndarray,
    horizon: float = 15.0,
) -> tuple[np.ndarray, np.ndarray, np.ndarray]:
    t_eval = np.linspace(0.0, horizon, 1501)

    def dynamics(t: float, x: np.ndarray) -> np.ndarray:
        return a_cl @ x + np.array([0.0, disturbance(t)], dtype=float)

    solution = solve_ivp(
        dynamics,
        (0.0, horizon),
        np.zeros(2, dtype=float),
        t_eval=t_eval,
        rtol=1e-8,
        atol=1e-10,
    )
    control = (k_gain @ solution.y).ravel()
    return solution.t, solution.y, control


def main() -> None:
    FIG_DIR.mkdir(parents=True, exist_ok=True)
    GEN_DIR.mkdir(parents=True, exist_ok=True)

    b_mat = np.array([[0.0], [1.0]], dtype=float)
    d_mat = np.array([[0.0], [1.0]], dtype=float)
    e_mat = np.diag([1.0, 0.4]).astype(float)
    q_mat = np.diag([6.0, 2.0]).astype(float)
    r_mat = np.array([[1.0]], dtype=float)

    a_nominal = uncertainty_matrix(0.0, 0.0)
    p_mat = linalg.solve_continuous_are(a_nominal, b_mat, q_mat, r_mat)
    k_gain = -np.linalg.solve(r_mat, b_mat.T @ p_mat)

    rho_1_grid = np.linspace(-0.6, 0.6, 81)
    rho_2_grid = np.linspace(-0.25, 0.25, 81)
    open_loop_alpha = np.zeros((rho_2_grid.size, rho_1_grid.size), dtype=float)
    closed_loop_alpha = np.zeros_like(open_loop_alpha)

    for i, rho_2 in enumerate(rho_2_grid):
        for j, rho_1 in enumerate(rho_1_grid):
            a_mat = uncertainty_matrix(rho_1, rho_2)
            a_cl = a_mat + b_mat @ k_gain
            open_loop_alpha[i, j] = np.max(np.linalg.eigvals(a_mat).real)
            closed_loop_alpha[i, j] = np.max(np.linalg.eigvals(a_cl).real)

    vertex_points = [
        ("V1", -0.6, -0.25),
        ("V2", -0.6, 0.25),
        ("V3", 0.6, -0.25),
        ("V4", 0.6, 0.25),
        ("Nominal", 0.0, 0.0),
    ]
    frequency_grid = np.logspace(-3, 3, 4000)
    vertex_metrics: list[dict[str, float | str]] = []
    for label, rho_1, rho_2 in vertex_points:
        a_mat = uncertainty_matrix(rho_1, rho_2)
        a_cl = a_mat + b_mat @ k_gain
        lyapunov_residual = a_cl.T @ p_mat + p_mat @ a_cl
        vertex_metrics.append(
            {
                "label": label,
                "rho_1": rho_1,
                "rho_2": rho_2,
                "open_loop_alpha": float(np.max(np.linalg.eigvals(a_mat).real)),
                "closed_loop_alpha": float(np.max(np.linalg.eigvals(a_cl).real)),
                "lyapunov_residual_alpha": float(
                    np.max(np.linalg.eigvals(lyapunov_residual).real)
                ),
                "hinf_estimate": estimate_hinf_norm(a_cl, d_mat, e_mat, frequency_grid),
            }
        )

    representative_cases = [
        ("rho1=-0.6, rho2=-0.25", -0.6, -0.25),
        ("rho1=0.0, rho2=0.0", 0.0, 0.0),
        ("rho1=0.6, rho2=0.25", 0.6, 0.25),
    ]
    response_data = []
    for label, rho_1, rho_2 in representative_cases:
        a_cl = uncertainty_matrix(rho_1, rho_2) + b_mat @ k_gain
        time, states, control = simulate_case(a_cl, k_gain)
        response_data.append(
            {
                "label": label,
                "rho_1": rho_1,
                "rho_2": rho_2,
                "time": time,
                "states": states,
                "control": control,
                "state_peak": float(np.max(np.abs(states))),
                "control_peak": float(np.max(np.abs(control))),
            }
        )

    summary = {
        "A_nominal": a_nominal.tolist(),
        "B": b_mat.tolist(),
        "D": d_mat.tolist(),
        "E": e_mat.tolist(),
        "K": k_gain.tolist(),
        "P": p_mat.tolist(),
        "uncertainty_box": {"rho_1": [-0.6, 0.6], "rho_2": [-0.25, 0.25]},
        "worst_closed_loop_alpha_on_grid": float(np.max(closed_loop_alpha)),
        "best_closed_loop_alpha_on_grid": float(np.min(closed_loop_alpha)),
        "worst_open_loop_alpha_on_grid": float(np.max(open_loop_alpha)),
        "best_open_loop_alpha_on_grid": float(np.min(open_loop_alpha)),
        "vertex_metrics": vertex_metrics,
        "representative_peaks": [
            {
                "label": item["label"],
                "state_peak": item["state_peak"],
                "control_peak": item["control_peak"],
            }
            for item in response_data
        ],
    }

    with (GEN_DIR / "summary.json").open("w", encoding="utf-8") as handle:
        json.dump(summary, handle, ensure_ascii=False, indent=2)

    with (GEN_DIR / "vertex_metrics.csv").open("w", encoding="utf-8", newline="") as handle:
        writer = csv.DictWriter(handle, fieldnames=list(vertex_metrics[0].keys()))
        writer.writeheader()
        writer.writerows(vertex_metrics)

    plt.rcParams.update(
        {
            "figure.figsize": (12, 4.8),
            "axes.grid": True,
            "grid.alpha": 0.25,
            "font.size": 11,
        }
    )

    fig, axes = plt.subplots(1, 2, constrained_layout=True)
    extent = [rho_1_grid[0], rho_1_grid[-1], rho_2_grid[0], rho_2_grid[-1]]
    image_1 = axes[0].imshow(
        open_loop_alpha,
        origin="lower",
        aspect="auto",
        extent=extent,
        cmap="magma",
    )
    axes[0].set_title("Open-loop spectral abscissa")
    axes[0].set_xlabel(r"$\rho_1$")
    axes[0].set_ylabel(r"$\rho_2$")
    fig.colorbar(image_1, ax=axes[0], shrink=0.85)

    image_2 = axes[1].imshow(
        closed_loop_alpha,
        origin="lower",
        aspect="auto",
        extent=extent,
        cmap="viridis",
    )
    axes[1].set_title("Closed-loop spectral abscissa")
    axes[1].set_xlabel(r"$\rho_1$")
    axes[1].set_ylabel(r"$\rho_2$")
    fig.colorbar(image_2, ax=axes[1], shrink=0.85)
    fig.savefig(FIG_DIR / "spectral_abscissa_scan.png", dpi=220)
    plt.close(fig)

    labels = [item["label"] for item in vertex_metrics]
    hinf_values = [float(item["hinf_estimate"]) for item in vertex_metrics]
    fig, ax = plt.subplots(figsize=(9, 4.8), constrained_layout=True)
    ax.bar(labels, hinf_values, color=["#456990", "#4d9078", "#f4d35e", "#ee964b", "#7d5ba6"])
    ax.axhline(0.5, color="#bc412b", linestyle="--", linewidth=1.6, label="gamma = 0.5")
    ax.set_ylabel("Estimated gain")
    ax.set_title(r"Estimated $H_\infty$ gain at interval vertices")
    ax.legend()
    fig.savefig(FIG_DIR / "vertex_hinf_estimates.png", dpi=220)
    plt.close(fig)

    fig, axes = plt.subplots(2, 1, figsize=(10.5, 6.4), sharex=True, constrained_layout=True)
    for item in response_data:
        axes[0].plot(item["time"], item["states"][0], label=f"{item['label']} : x1")
        axes[1].plot(item["time"], item["states"][1], label=f"{item['label']} : x2")
    axes[0].set_ylabel("x1(t)")
    axes[1].set_ylabel("x2(t)")
    axes[1].set_xlabel("Time (s)")
    axes[0].set_title("Closed-loop states under decaying disturbance")
    axes[0].legend(ncol=3, fontsize=8)
    axes[1].legend(ncol=3, fontsize=8)
    fig.savefig(FIG_DIR / "state_response_under_disturbance.png", dpi=220)
    plt.close(fig)

    fig, ax = plt.subplots(figsize=(10.5, 4.6), constrained_layout=True)
    for item in response_data:
        ax.plot(item["time"], item["control"], linewidth=1.8, label=item["label"])
    sample_time = response_data[0]["time"]
    ax.plot(sample_time, disturbance(sample_time), color="#222222", linestyle="--", linewidth=1.2, label="w(t)")
    ax.set_xlabel("Time (s)")
    ax.set_ylabel("Signal value")
    ax.set_title("Control input and disturbance")
    ax.legend(ncol=4, fontsize=8)
    fig.savefig(FIG_DIR / "control_response_under_disturbance.png", dpi=220)
    plt.close(fig)

    print("Robust control reproduction completed.")
    print(f"K = {k_gain}")
    print(f"Worst closed-loop spectral abscissa on grid: {np.max(closed_loop_alpha):.6f}")
    print(f"Worst estimated vertex Hinf gain: {max(hinf_values):.6f}")


if __name__ == "__main__":
    main()
